homeassistant/components/device_tracker/icloud.py

```diff
@@ -1,100 +1,455 @@
+                self.update_device(devicename)
+
+    def determine_interval(self, devicename, latitude, longitude, battery):
+        """Calculate new interval."""
+        # pylint: disable=too-many-branches
+        distancefromhome = None
+        zone_state = self.hass.states.get('zone.home')
+        zone_state_lat = zone_state.attributes['latitude']
+        zone_state_long = zone_state.attributes['longitude']
+        distancefromhome = distance(latitude, longitude, zone_state_lat,
+                                    zone_state_long)
+        distancefromhome = round(distancefromhome / 1000, 1)
+
+        currentzone = active_zone(self.hass, latitude, longitude)
+
+        if ((currentzone is not None and
+             currentzone == self._overridestates.get(devicename)) or
+                (currentzone is None and
+                 self._overridestates.get(devicename) == 'away')):
+            return
+
+        self._overridestates[devicename] = None
+
+        if currentzone is not None:
+            self._intervals[devicename] = 30
+            return
+
+        if distancefromhome is None:
+            return
+        if distancefromhome > 100:
+            self._intervals[devicename] = round(distancefromhome, 0)
+            gtt = self.googletraveltime.get(devicename)
+            if gtt is not None:
+                gttstate = self.hass.states.get(gtt)
+                if gttstate is not None:
+                    self._intervals[devicename] = round(
+                        float(gttstate.state) - 10, 0)
+        elif distancefromhome > 50:
+            self._intervals[devicename] = 30
+        elif distancefromhome > 25:
+            self._intervals[devicename] = 15
+        elif distancefromhome > 10:
+            self._intervals[devicename] = 5
+        else:
+            self._intervals[devicename] = 1
+        if battery is not None and battery <= 33 and distancefromhome > 3:
+            self._intervals[devicename] = self._intervals[devicename] * 2
+
+    def update_device(self, devicename):
+        """Update the device_tracker entity."""
+        # pylint: disable=too-many-branches,too-many-statements
+        # pylint: disable=too-many-nested-blocks,too-many-locals
+        from pyicloud.exceptions import PyiCloudNoDevicesException
+
+        entity = self.hass.states.get(ENTITY_ID_FORMAT.format(devicename))
+        if entity is None and devicename in self.seen_devices:
+            return
+        attrs = {}
+        kwargs = {}
+        gtt = self.googletraveltime.get(devicename)
+        if gtt is not None:
+            gttstate = self.hass.states.get(gtt)
+            if gttstate is not None:
+                if 'origin_addresses' in gttstate.attributes:
+                    duration = gttstate.state
+                    origin = gttstate.attributes['origin_addresses']
+                    attrs[ATTR_GMTT_DURATION] = duration
+                    attrs[ATTR_GMTT_ORIGIN] = origin
```

Comment:
Instead of testing gtt or gttstate, rather wrap this section in a try...except, so if someone points to a sensor that is not a google travel time it wont blow up.

So line 358 to 366 would change to

```python
try:
    gtt = self.googletraveltime[devicename]
    gttstate = self.hass.states.get(gtt)
    attrs[ATTR_GMTT_ORIGIN] = gttstate.attributes['origin_addresses']
    attrs[ATTR_GMTT_DURATION] = gttstate.state
except AttributeError, KeyError:
  pass
```